retry.go

```diff
@@ -0,0 +1,70 @@
+package rueidis
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"github.com/redis/rueidis/internal/util"
+)
+
+const (
+	defaultMaxRetryDelay = 1 * time.Second
+	maxAttemptsShift     = 63 // Avoid excessive shifts
+)
+
+// RetryDelay returns the delay that should be used before retrying the
+// attempt. Will return negative delay if the delay could not be determined or do not retry.
+type RetryDelay func(attempts int, err error) time.Duration
+
+// defaultRetryDelay delays the next retry exponentially without considering the error.
+// max delay is 1 second.
+func defaultRetryDelay(attempts int, _ error) time.Duration {
+	base := util.FastRandFloat64()
+	backoff := uint64(1 << uint64(math.Min(float64(attempts), maxAttemptsShift)))
+	jitter := base * float64(backoff) * float64(time.Millisecond)
+	return time.Duration(math.Min(jitter, float64(defaultMaxRetryDelay)))
+}
+
+// retryableCheckFunc is a function that checks if the command should be retried.
+// Returns true if the command should be retried.
+type retryableCheckFunc func() bool
+
+type retryHandler interface {
+	// WaitUntilNextRetry waits until the next retry should be attempted.
+	// Returns false without error immediately if the command should not be retried.
+	// Returns false and an error if waiting for the next retry was interrupted.
+	// Returns true after the delay if the command should be retried.
+	WaitUntilNextRetry(ctx context.Context, isRetryable retryableCheckFunc, attempts int, err error) (bool, error)
+}
+
+type retryer struct {
+	RetryDelay RetryDelay
+}
+
+func newRetryer(retryDelay RetryDelay) *retryer {
+	return &retryer{RetryDelay: retryDelay}
+}
+
+func (r *retryer) WaitUntilNextRetry(
+	ctx context.Context, isRetryable retryableCheckFunc, attempts int, err error,
+) (bool, error) {
+	if !isRetryable() {
```

Comment:
I don't like the `isRetryable retryableCheckFunc` to be passed in here. I think `isRetryable` should be checked before calling the `WaitUntilNextRetry`.

Reply:
[9e884c3](https://github.com/redis/rueidis/pull/643/commits/9e884c395cb0aa8c382e10c1dd92a514130c1efb). OK.